File: MAPageViewController/MAPageViewController.m
//
//  MAPageViewController.m
//  MAPageViewController
//
//  Created by Mike on 7/24/14.
//  Copyright (c) 2014 Mike Amaral. All rights reserved.
//

#import "MAPageViewController.h"

static const CGFloat kMAPageControlHeight = 40;

@interface MAPageViewController ()

@end

@implementation MAPageViewController

- (instancetype)initWithViewControllers:(NSArray *)viewControllers {
    self = [super init];
    
    _viewControllers = viewControllers;
    
    return self;
}


#pragma mark - View life cycle

- (void)viewDidLoad {
    [super viewDidLoad];
    self.edgesForExtendedLayout = UIRectEdgeNone;
    
    [self setupPageViewController];
}

- (void)setupPageViewController {
    _pageViewController = [[UIPageViewController alloc] initWithTransitionStyle:UIPageViewControllerTransitionStyleScroll navigationOrientation:UIPageViewControllerNavigationOrientationHorizontal options:nil];
    _pageViewController.delegate = self;
    _pageViewController.dataSource = self;
    _pageViewController.view.frame = self.view.frame;
    [_pageViewController setViewControllers:@[_viewControllers[0]] direction:UIPageViewControllerNavigationDirectionForward animated:YES completion:nil];
    [self addChildViewController:_pageViewController];
    [self.view addSubview:_pageViewController.view];
    [_pageViewController didMoveToParentViewController:self];
    
    _pageControl = [[UIPageControl alloc] initWithFrame:CGRectMake(0, CGRectGetMaxY(self.view.bounds) - kMAPageControlHeight, self.view.bounds.size.width, kMAPageControlHeight)];
    _pageControl.numberOfPages = _viewControllers.count;
    _pageControl.autoresizingMask = UIViewAutoresizingFlexibleWidth + UIViewAutoresizingFlexibleTopMargin;
    [self.view addSubview:_pageControl];
}

- (NSInteger)indexOfViewController:(UIViewController *)viewController {
    NSInteger theIndex = 0;
    for (NSInteger index = 0; index < _viewControllers.count; index++) {
        if (_viewControllers[index] == viewController) {
            theIndex = index;
        }
    }
    return theIndex;
}


#pragma mark - Page view controller delegate

- (void)pageViewController:(UIPageViewController *)pageViewController didFinishAnimating:(BOOL)finished previousViewControllers:(NSArray *)previousViewControllers transitionCompleted:(BOOL)completed {
    if (!completed) {
        return;
    }
    
    UIViewController *newViewController = pageViewController.viewControllers[0];
    _pageControl.currentPage = [self indexOfViewController:newViewController];
}


#pragma mark - Page view controller data source

- (UIViewController *)pageViewController:(UIPageViewController *)pageViewController viewControllerBeforeViewController:(UIViewController *)viewController {
    NSInteger indexOfCurrentViewController = [self indexOfViewController:viewController];
    return indexOfCurrentViewController > 0 ? _viewControllers[indexOfCurrentViewController - 1] : nil;
}

- (UIViewController *)pageViewController:(UIPageViewController *)pageViewController viewControllerAfterViewController:(UIViewController *)viewController {
    NSInteger indexOfCurrentViewController = [self indexOfViewController:viewController];
    return indexOfCurrentViewController < _viewControllers.count - 1 ? _viewControllers[indexOfCurrentViewController + 1] : nil;
}

@end

File: MAPageViewController/MAPageViewController.swift
//
//  MAPageViewController.swift
//  MAPageViewController
//
//  Created by Mike on 6/19/14.
//  Copyright (c) 2014 Mike Amaral. All rights reserved.
//

import UIKit

class MAPageViewController: UIViewController, UIPageViewControllerDataSource, UIPageViewControllerDelegate {
    let pageViewController: UIPageViewController = UIPageViewController(transitionStyle: UIPageViewControllerTransitionStyle.Scroll, navigationOrientation: UIPageViewControllerNavigationOrientation.Horizontal, options: nil)
    let pageControl: UIPageControl = UIPageControl()
    let pageControlHeight: CGFloat = 40
    var viewControllers: UIViewController[] = []

    init(viewControllers: UIViewController[]) {
        self.viewControllers = viewControllers
        super.init(nibName: nil, bundle: nil)
    }
    
    init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: NSBundle?) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        self.edgesForExtendedLayout = .None
        setupPageViewController()
    }
    
    func setupPageViewController() {
        pageViewController.delegate = self
        pageViewController.dataSource = self;
        pageViewController.view.frame = self.view.frame;
        pageViewController.setViewControllers([viewControllers[0]], direction: UIPageViewControllerNavigationDirection.Forward, animated: false, completion: nil)
        self.addChildViewController(pageViewController)
        self.view.addSubview(pageViewController.view)
        pageViewController.didMoveToParentViewController(self)
        
        pageControl.frame = CGRectMake(0, CGRectGetMaxY(self.view.frame) - pageControlHeight, self.view.bounds.size.width, pageControlHeight)
        pageControl.numberOfPages = viewControllers.count;
        pageControl.autoresizingMask = .FlexibleWidth | .FlexibleTopMargin
        self.view.addSubview(pageControl)
    }
    
    func indexOfViewController(viewController: UIViewController) -> Int {
        var indexOfVC: Int = 0
        for (index, element) in enumerate(viewControllers) {
            if element == viewController {
                indexOfVC = index
                break
            }
        }
        return indexOfVC
    }
    
    // PRAGMA: page view controller data source
    
    func pageViewController(pageViewController: UIPageViewController!, viewControllerAfterViewController viewController: UIViewController!) -> UIViewController! {
        let indexOfCurrentVC = indexOfViewController(viewController)
        return indexOfCurrentVC < viewControllers.count - 1 ? viewControllers[indexOfCurrentVC + 1] : nil
    }
    
    func pageViewController(pageViewController: UIPageViewController!, viewControllerBeforeViewController viewController: UIViewController!) -> UIViewController!  {
        viewControllers.endIndex
        let indexOfCurrentVC = indexOfViewController(viewController)
        return indexOfCurrentVC > 0 ? viewControllers[indexOfCurrentVC - 1] : nil
    }
    
    // PRAGMA: page view controller delegate
    
    func pageViewController(pageViewController: UIPageViewController!, didFinishAnimating finished: Bool, previousViewControllers: AnyObject[]!, transitionCompleted completed: Bool) {
        if !completed {
            return
        }
        
        let newViewController = pageViewController.viewControllers[0] as UIViewController
        pageControl.currentPage = indexOfViewController(newViewController)
    }
    
    

}
